src/clj/rems/db/entitlements.clj

```diff
@@ -63,52 +64,71 @@
 
 (defn update-entitlements-for
+  "If the given application is approved, licenses accepted etc. add an entitlement to the db
+  and call the entitlement REST callback (if defined). Likewise if a resource is removed, member left etc.
+  then we end the entitlement and call the REST callback."
   [application]
-  (add-entitlements-for application)
-  (end-entitlements-for application))
+  (when (contains? #{:application.state/approved :application.state/closed}
+                   (:application/state application))
+    (let [application-id (:application/id application)
+          current-members (set (concat (map :userid (:application/members application))
+                                       [(:application/applicant application)]))
+          past-members (set (map :userid (:application/past-members application)))
+          application-state (:application/state application)
+          application-resources (->> application
+                                     :application/resources
+                                     (map :resource/id)
+                                     set)
+          application-entitlements (get-entitlements-by-user application-id)
+          is-entitled? (fn [userid application resource-id]
+                         (and (= :application.state/approved application-state)
+                              (contains? current-members userid)
+                              (accepted-licenses? application userid)
+                              (contains? application-resources resource-id)))
+          entitlements-by-user (fn [userid] (or (application-entitlements userid) #{}))
+          entitlements-to-add (->> (for [userid (union current-members past-members)
+                                         :let [current-resource-ids (entitlements-by-user userid)]
+                                         resource-id application-resources
+                                         :when (is-entitled? userid application resource-id)
+                                         :when (not (contains? current-resource-ids resource-id))]
+                                     {userid #{resource-id}})
+                                   (apply merge-with union))
+          entitlements-to-remove (->> (for [userid (union current-members past-members)
+                                            :let [resource-ids (entitlements-by-user userid)]
+                                            resource-id resource-ids
+                                            :when (not (is-entitled? userid application resource-id))]
+                                        {userid #{resource-id}})
+                                      (apply merge-with union))
+          members-to-update (set
+                             (concat (keys entitlements-to-add)
+                                     (keys entitlements-to-remove)))]
+      (when (seq members-to-update)
```

Comment:
hmm if you split out the computation of `entitlements-to-{add,remove}` you could write unit tests for it

Reply:
I can't see an easy way to split them out that would not duplicate the fetching logic. I'll leave this for later PRs.